### PR TITLE
RPC: Add parameter to addmultisigaddress / createmultisig to sort public keys

### PR DESCRIPTION
**Pull request by afk11**
I figured it may be useful for these RPC methods to allow sorting public keys (BIP67) The PR adds a new boolean to createmultisig / addmultisigaddress at the end of their parameter list. By default, this is set to false to avoid a BC break. 

I added a RPC test file `sort_multisig.py` for testing createmultisig. Tests for addmultisigaddress went in `wallet-accounts.py`.

Note: Code to check whether sorting is desired had to be replicated in both RPC methods (not in _createmultisig_redeemScript) because addmultisigaddress already takes a parameter at position 3.

### CODE REVIEW

**Comment by dcousens** [comment]
concept ACK

**Comment by laanwj** [comment]
Concept ACK, although I really don't like multiple-optional-boolean APIs. Wish we switched to named arguments any day.

One nit: the RPC help should mention BIP67 by name.

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by afk11** [comment]
@MarcoFalke thanks, done. @laanwj I should have mentioned, nits addressed. 

One travis run failed due to the compactblocks RPC test.

**Comment by ryanofsky** [comment]
I can't see anything on travis right now  (503 errors), but the compactblocks error is probably just the spurious https://github.com/bitcoin/bitcoin/issues/8842 / https://github.com/bitcoin/bitcoin/pull/9058 failures.

File: qa/rpc-tests/sort_multisig.py
Comment (MarcoFalke) [inline qa/rpc-tests/sort_multisig.py]: Nit: A single node should be enough?

**Comment by MarcoFalke** [comment]
Concept ACK 7d7a64726991ff087cb8125e0c7277173a688dc7

File: src/rpc/misc.cpp
Comment (luke-jr) [inline src/rpc/misc.cpp]: Is it a string or a boolean?

Comment (afk11) [inline src/rpc/misc.cpp]: It should be a boolean. Just observed they aren't usually quoted in RPC output, fixing now.

**Comment by afk11** [comment]
I probably shouldn't have squashed @MarcoFalke, I'm sorry for rebasing out the commit you reviewed. The only thing to change this time was the removal of "'s from the RPC help message.

File: src/script/standard.cpp
Comment (luke-jr) [inline src/script/standard.cpp]: Wouldn't it be better to `reserve` and then `emplace_back`?

File: src/script/standard.cpp
Comment (luke-jr) [inline src/script/standard.cpp]: I *think* this should do what BIP 67 requires, but someone more familiar with C++ and its locale support (or lack thereof) should probably confirm.

File: src/script/standard.h
Comment (luke-jr) [inline src/script/standard.h]: Maybe default fSorted to false here rather than modify all the tests?

File: src/wallet/rpcwallet.cpp
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: As before, rather turn param 3 into an options Object.

**Review by luke-jr — CHANGES_REQUESTED** [review]
Code looks reasonably correct, just a few nits. Did not verify tests.

**Comment by luke-jr** [comment]
Should update `doc/bips.md` also.

**Comment by luke-jr** [comment]
I rebased and addressed all the nits; pushed this to luke-jr/sort-multisigs

@afk11 Are you still maintaining this? Can you pull my changes?

```
git checkout sort-multisigs
git fetch git://github.com/luke-jr/bitcoin sort-multisigs
git reset --hard FETCH_HEAD
git push ...
```

**Comment by afk11** [comment]
Sorry, yep I can pull these! 

I wanted to wait until named parameters was merged before hand, so I could avoid adding a positional parameter before the accounts parameters were changed 

I'll look at this in the next day or so (away from internet atm) wanna finish this up

**Comment by afk11** [comment]
Merged commits and rebased. Apologies for the delay!

**Comment by afk11** [comment]
The Apple build failed because the job time exceeded the maximum :/

**Comment by afk11** [comment]
Rebased

**Comment by afk11** [comment]
Rebased

**Comment by luke-jr** [comment]
The description for d70583a2f9ae8bb4bda2e6224e699ac2cbb583ee is no longer correct.

**Comment by afk11** [comment]
@luke-jr Oops, yeah the rebase mightn't have been the easiest. So re-add the contents of setup_network?

**Comment by jnewbery** [comment]
Perhaps I'm missing something, but I don't see the need for this. The `addmultisigaddress` RPC creates the multisig script with the keys in the order provided. Why not just have the client provide keys in sorted order if you want the script to be BIP-67 compliant?

It doesn't look like this PR is enforcing that the provided keys are compressed, so even with this PR, there are still expectations placed on the client.

**Comment by afk11** [comment]
I think if developers are already committing to using the RPC to make a multisig script, making it easier to produce the same representation is more important than not.

You are correct the PR as it stands doesn't validate it.. fixing this now.

**Comment by jnewbery** [comment]
I'm still a weak concept NACK for this. I don't agree that we should add complexity to the server when the same outcome can be achieved by simply running a `sort()` function on the client before calling the RPC. Sometimes there's good reason to add that complexity to the server - see for example #9991 which adds a filter to save significantly on bandwidth and server resources. In this case I don't see the benefit.

Sorry if that sounds negative - I think there needs to be some bar for adding new RPCs and arguments to avoid bloat.

However, if I'm wrong and there's widespread consensus that this is useful functionality and should be merged, can I at least ask that you use named arguments instead of an Options object? There's really no need for Options objects in RPC calls since named args were added in #8811.

**Comment by afk11** [comment]
Both RPC methods take an options object for this (sorry, the PR description wasn't updated with this)
https://github.com/bitcoin/bitcoin/pull/8751/files#diff-ad6efdc354b57bd1fa29fc3abb6e2872R353
https://github.com/bitcoin/bitcoin/pull/8751/files#diff-df7d84ff2f53fcb2a0dc15a3a51e55ceR1050 

I appreciate where you are coming from and agree that most people can probably sort themselves, but they could also build a multisig script out of the keys and m/n. It's been a while since I've even used the RPC, but remember well the time when I didn't have a bitcoin library to do it all. 

I think it's worth including since once they continue using the flag, requests which mistakenly use the wrong order will reproduce the same redeem script (instead of always having a stateful order of public keys), and likewise with libraries that support it.

**Comment by ipoptika** [comment]
https://minecloud.io/?in=ipoptika

File: src/script/standard.cpp
Comment (luke-jr) [inline src/script/standard.cpp]: We're not using `BOOST_FOREACH` anymore I think. (Also below)

**Comment by luke-jr** [comment]
Rebased and squashed a bit.

```
git checkout sort-multisigs && git fetch git://github.com/luke-jr/bitcoin sort-multisigs && git reset --hard FETCH_HEAD && git push ...
```

**Comment by TheBlueMatt** [comment]
Concept ACK. Care to rebase?

**Comment by afk11** [comment]
Rebased, sorry for the delay. Updated to check that keys are compressed before allowing sorting, and added more tests for this. 

Updated the docs/bips.md document to mention 0.15.1 instead of 0.15.0 (let me know whatever's best for this)

**Comment by TheBlueMatt** [comment]
Hmm, hate to reopen it, but now that we do actually have named arguments, could you rever to just adding a new boolean argument? options objects are just redundant now, and having options alias account in addmultisigaddress is just gross. Everything else seems fine at first glance.

@jnewbery I'd generally agree with you, but, at least in principal, I think BIP67 is worth it.

**Comment by afk11** [comment]
@TheBlueMatt that's fine, revised the PR now. 

I missed the boat again for v0.15.1, suggestions for a release to mention in bips.md?

**Comment by luke-jr** [comment]
> Hmm, hate to reopen it, but now that we do actually have named arguments, could you rever to just adding a new boolean argument? options objects are just redundant now, and having options alias account in addmultisigaddress is just gross. Everything else seems fine at first glance.

Strongly disagree. Named arguments is not a reason to have a terrible positional arguments API. Uncommon options should go through an options argument when positional arguments are used.

The account alias is merely for backward compatibility.

**Comment by laanwj** [comment]
Closing and putting "up for grabs" label

**Pull request closed without merging**
